Specify zk chroot path failed when initialize cluster metadata

When trying to initialize cluster metadata via the following shell:
```
bin/pulsar initialize-cluster-metadata \
  --cluster pulsar-cluster-test\
  --zookeeper hdp1:2181/kobe \
  --configuration-store hdp1:2181/kobe \
  --web-service-url http://hdp1:8080 \
  --web-service-url-tls https://hdp1:8443 \
  --broker-service-url pulsar://hdp1:6650 \
  --broker-service-url-tls pulsar+ssl://hdp1:6651
```
and the config in `bookkeeper.conf` likes this:
```
zkServers=hdp1:2181
zkLedgersRootPath=/kobe/ledgers
```
It doesn't work well as below:
```
Exception in thread "main" java.util.concurrent.ExecutionException: KeeperErrorCode = NoNode for /kobe/ledgers
	at org.apache.bookkeeper.meta.MetadataDrivers.runFunctionWithMetadataBookieDriver(MetadataDrivers.java:378)
	at org.apache.bookkeeper.client.BookKeeperAdmin.format(BookKeeperAdmin.java:1150)
	at org.apache.pulsar.PulsarClusterMetadataSetup.main(PulsarClusterMetadataSetup.java:152)
Caused by: org.apache.zookeeper.KeeperException$NoNodeException: KeeperErrorCode = NoNode for /kobe/ledgers
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:114)
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:54)
	at org.apache.zookeeper.ZooKeeper.create(ZooKeeper.java:792)
	at org.apache.bookkeeper.zookeeper.ZooKeeperClient.access$1901(ZooKeeperClient.java:70)
	at org.apache.bookkeeper.zookeeper.ZooKeeperClient$9.call(ZooKeeperClient.java:711)
	at org.apache.bookkeeper.zookeeper.ZooKeeperClient$9.call(ZooKeeperClient.java:705)
	at org.apache.bookkeeper.zookeeper.ZooWorker.syncCallWithRetries(ZooWorker.java:140)
	at org.apache.bookkeeper.zookeeper.ZooKeeperClient.create(ZooKeeperClient.java:705)
	at org.apache.bookkeeper.discover.ZKRegistrationManager.prepareFormat(ZKRegistrationManager.java:403)
	at org.apache.bookkeeper.client.BookKeeperAdmin.lambda$format$2(BookKeeperAdmin.java:1152)
	at org.apache.bookkeeper.meta.MetadataDrivers.runFunctionWithMetadataBookieDriver(MetadataDrivers.java:373)
```
As a workaround, you should `create /kobe` through the zk shell at first, it is better to avoid to create the prefix path manually.

--- a/pulsar-broker/src/main/java/org/apache/pulsar/PulsarClusterMetadataSetup.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/PulsarClusterMetadataSetup.java
@@ -270,7 +270,7 @@ public class PulsarClusterMetadataSetup {
         log.info("Cluster metadata for '{}' setup correctly", arguments.cluster);
     }
 
-    private static ZooKeeper initZk(String connection, int sessionTimeout) throws Exception {
+    public static ZooKeeper initZk(String connection, int sessionTimeout) throws Exception {
         ZooKeeperClientFactory zkfactory = new ZookeeperClientFactoryImpl();
         int chrootIndex = connection.indexOf("/");
         if (chrootIndex > 0) {

--- a/pulsar-broker/src/main/java/org/apache/pulsar/PulsarClusterMetadataSetup.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/PulsarClusterMetadataSetup.java
@@ -31,7 +31,6 @@ import java.util.List;
 import org.apache.bookkeeper.client.BookKeeperAdmin;
 import org.apache.bookkeeper.common.net.ServiceURI;
 import org.apache.bookkeeper.conf.ServerConfiguration;
-import org.apache.bookkeeper.meta.HierarchicalLedgerManagerFactory;
 import org.apache.bookkeeper.stream.storage.api.cluster.ClusterInitializer;
 import org.apache.bookkeeper.stream.storage.impl.cluster.ZkClusterInitializer;
 import org.apache.bookkeeper.util.ZkUtils;
@@ -93,7 +92,7 @@ public class PulsarClusterMetadataSetup {
         private String globalZookeeper;
 
         @Parameter(names = { "-cs",
-            "--configuration-store" }, description = "Configuration Store connection string", required = false)
+            "--configuration-store" }, description = "Configuration Store connection string", required = true)
         private String configurationStore;
 
         @Parameter(names = {
@@ -136,13 +135,11 @@ public class PulsarClusterMetadataSetup {
             arguments.configurationStore = arguments.globalZookeeper;
         }
 
-        log.info("Setting up cluster {} with zk={} configuration-store ={}", arguments.cluster, arguments.zookeeper,
+        log.info("Setting up cluster {} with zk={} configuration-store={}", arguments.cluster, arguments.zookeeper,
                 arguments.configurationStore);
-        ZooKeeperClientFactory zkfactory = new ZookeeperClientFactoryImpl();
-        ZooKeeper localZk = zkfactory.create(
-            arguments.zookeeper, SessionType.ReadWrite, arguments.zkSessionTimeoutMillis).get();
-        ZooKeeper configStoreZk = zkfactory.create(
-            arguments.configurationStore, SessionType.ReadWrite, arguments.zkSessionTimeoutMillis).get();
+
+        ZooKeeper localZk = initZk(arguments.zookeeper, arguments.zkSessionTimeoutMillis);
+        ZooKeeper configStoreZk = initZk(arguments.configurationStore, arguments.zkSessionTimeoutMillis);
 
         // Format BookKeeper ledger storage metadata
         ServerConfiguration bkConf = new ServerConfiguration();
@@ -271,6 +268,29 @@ public class PulsarClusterMetadataSetup {
         }
 
         log.info("Cluster metadata for '{}' setup correctly", arguments.cluster);
+    }
+
+    private static ZooKeeper initZk(String connection, int sessionTimeout) throws Exception {
+        ZooKeeperClientFactory zkfactory = new ZookeeperClientFactoryImpl();
+        int chrootIndex = connection.indexOf("/");
+        if (chrootIndex > 0) {
+            String chrootPath = connection.substring(chrootIndex);
+            String zkConnectForChrootCreation = connection.substring(0, chrootIndex);
+            ZooKeeper chrootZk = zkfactory.create(
+                zkConnectForChrootCreation, SessionType.ReadWrite, sessionTimeout).get();
+            if (chrootZk.exists(chrootPath, false) == null) {
+                try {
+                    ZkUtils.createFullPathOptimistic(chrootZk, chrootPath, new byte[0], ZooDefs.Ids.OPEN_ACL_UNSAFE,
+                        CreateMode.PERSISTENT);
+                } catch (NodeExistsException e) {
+                    // Ignore
+                }
+                log.info("Created zookeeper chroot path {} successfully", chrootPath);
+            }
+            chrootZk.close();
+        }
+        ZooKeeper zkConnect = zkfactory.create(connection, SessionType.ReadWrite, sessionTimeout).get();
+        return zkConnect;
     }
 
     private static BundlesData getBundles(int numBundles) {

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/zookeeper/ZooKeeperClientAspectJTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/zookeeper/ZooKeeperClientAspectJTest.java
@@ -34,6 +34,7 @@ import java.util.concurrent.atomic.AtomicInteger;
 
 import org.apache.bookkeeper.common.util.OrderedScheduler;
 import org.apache.bookkeeper.test.PortManager;
+import org.apache.pulsar.PulsarClusterMetadataSetup;
 import org.apache.pulsar.broker.PulsarService;
 import org.apache.pulsar.broker.service.BrokerService;
 import org.apache.pulsar.broker.service.BrokerTestBase;
@@ -46,6 +47,7 @@ import org.apache.pulsar.zookeeper.LocalBookkeeperEnsemble;
 import org.apache.pulsar.zookeeper.ZooKeeperClientFactory;
 import org.apache.pulsar.zookeeper.ZooKeeperClientFactory.SessionType;
 import org.apache.pulsar.zookeeper.ZookeeperBkClientFactoryImpl;
+import org.apache.pulsar.zookeeper.ZookeeperClientFactoryImpl;
 import org.apache.zookeeper.CreateMode;
 import org.apache.zookeeper.ZooDefs;
 import org.apache.zookeeper.ZooKeeper;
@@ -94,6 +96,30 @@ public class ZooKeeperClientAspectJTest {
             }
 
             executor.shutdown();
+        }
+    }
+
+    @Test
+    public void testInitZk() throws Exception {
+        try {
+            ZooKeeperClientFactory zkfactory = new ZookeeperClientFactoryImpl();
+            CompletableFuture<ZooKeeper> zkFuture = zkfactory.create("127.0.0.1:" + LOCAL_ZOOKEEPER_PORT, SessionType.ReadWrite,
+                (int) ZOOKEEPER_SESSION_TIMEOUT_MILLIS);
+            localZkc = zkFuture.get(ZOOKEEPER_SESSION_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+            assertTrue(localZkc.getState().isConnected());
+            assertNotEquals(localZkc.getState(), States.CONNECTEDREADONLY);
+
+            String connection = "127.0.0.1:" + LOCAL_ZOOKEEPER_PORT + "/prefix";
+            ZooKeeper chrootZkc = PulsarClusterMetadataSetup.initZk(connection, (int) ZOOKEEPER_SESSION_TIMEOUT_MILLIS);
+            assertTrue(chrootZkc.getState().isConnected());
+            assertNotEquals(chrootZkc.getState(), States.CONNECTEDREADONLY);
+            chrootZkc.close();
+
+            assertTrue(localZkc.exists("/prefix", false) != null);
+        } finally {
+            if (localZkc != null) {
+                localZkc.close();
+            }
         }
     }
 

